src/core/thread/mle_router.cpp

```diff
@@ -1106,6 +1107,13 @@ ThreadError MleRouter::HandleAdvertisement(const Message &aMessage, const Ip6::M
                      leaderData.GetWeighting(), peerParitionId,
                      mLeaderData.GetWeighting(), mLeaderData.GetPartitionId());
 
+        if (((mDeviceMode & ModeTlv::kModeFFD) == 0) && (memcmp(&mParent.mMacAddr, &macAddr, sizeof(router->mMacAddr)) == 0) &&
+            (mParent.mValid.mRloc16 == sourceAddress.GetRloc16()))
+        {
+            SetLeaderData(leaderData.GetPartitionId(), leaderData.GetWeighting(), leaderData.GetLeaderRouterId());
+            ExitNow();
+        }
```

Comment:
This code block should probably live in `mle.cpp`, since it is required for all Thread devices.